chiselFrontend/src/main/scala/chisel3/RawModule.scala

```diff
@@ -134,6 +134,14 @@ abstract class RawModule(implicit moduleCompileOptions: CompileOptions)
   }
 }
 
+trait RequireAsyncReset extends MultiIOModule {
+  override private[chisel3] def mkReset: AsyncReset = AsyncReset()
```

Comment:
Should this be `final`?

Or: what should happen for `new Foo with RequireAsyncReset with RequireSyncReset`?

Reply:
At the moment, you'll get this error:
```
error] /Users/jack/work/chisel3-control-reset-type/src/main/scala/Jack.scala:7:7: overriding method mkReset in trait RequireAsyncReset of type => chisel3.AsyncReset;
[error]  method mkReset in trait RequireSyncReset of type => chisel3.Bool has incompatible type
[error] class MyModule extends Module with MyTrait with RequireAsyncReset with RequireSyncReset {
[error]       ^
```
If I make them both `final`, you get
```
[error] /Users/jack/work/chisel3-control-reset-type/src/main/scala/Jack.scala:7:7: overriding method mkReset in trait RequireAsyncReset of type => chisel3.AsyncReset;
[error]  method mkReset in trait RequireSyncReset of type => chisel3.Bool cannot override final member
[error] class MyModule extends Module with MyTrait with RequireAsyncReset with RequireSyncReset {
[error]       ^
```
I would argue the formal (ie. without `final`) is actually marginally better

Reply:
Ah, I see. You get mutual exclusion through the type system. I think that's fine.